typescript/cosmos-sdk/src/index.ts

```diff
@@ -0,0 +1,603 @@
+    return new HyperlaneModuleClient(client, options);
+  }
+
+  public async simulate(
+    signerAddress: string,
+    signerPubKey: Pubkey,
+    messages: any[],
+    memo: string | undefined,
+  ): Promise<number> {
+    const queryClient = this.getQueryClient()!;
+
+    const { sequence } = await this.getSequence(signerAddress);
+    const { gasInfo } = await queryClient.tx.simulate(
+      messages,
+      memo,
+      signerPubKey,
+      sequence,
+    );
+    return Uint53.fromString(gasInfo?.gasUsed.toString() ?? '0').toNumber();
+  }
+}
+
+export class SigningHyperlaneModuleClient extends SigningStargateClient {
+  public query: HyperlaneQueryClient;
+  public account: AccountData;
+  public readonly GAS_MULTIPLIER = 1.6;
+
+  protected constructor(
+    cometClient: CometClient,
+    signer: OfflineSigner,
+    account: AccountData,
+    options: SigningStargateClientOptions,
+  ) {
+    // register all the custom amino tx types
+    const aminoTypes = Object.values(REGISTRY)
+      .filter((r) => !!r.amino.type) // filter out tx responses which have no amino type
+      .reduce(
+        (types, { proto, amino }) => ({
+          ...types,
+          [proto.type]: {
+            aminoType: amino.type,
+            toAmino: (amino.converter as any)?.toJSON ?? proto.converter.toJSON,
+            fromAmino:
+              (amino.converter as any)?.fromJSON ?? proto.converter.fromJSON,
+          },
+        }),
+        {},
+      );
+
+    super(cometClient, signer, {
+      ...options,
+      aminoTypes: new AminoTypes({
+        ...options.aminoTypes,
+        ...aminoTypes,
+      }),
+    });
+
+    this.query = QueryClient.withExtensions(
+      cometClient,
+      setupBankExtension,
+      setupCoreExtension,
+      setupInterchainSecurityExtension,
+      setupPostDispatchExtension,
+      setupWarpExtension,
+    );
+
+    // register all the custom tx types
+    Object.values(REGISTRY).forEach(({ proto }) => {
+      this.registry.register(proto.type, proto.converter);
+    });
+
+    this.account = account;
+  }
+
+  static async connectWithSigner(
+    endpoint: string | HttpEndpoint,
+    signer: OfflineSigner,
+    options: SigningStargateClientOptions = {},
+  ): Promise<SigningHyperlaneModuleClient> {
+    const client = await connectComet(endpoint);
+    const [account] = await signer.getAccounts();
+    return new SigningHyperlaneModuleClient(client, signer, account, options);
+  }
+
+  static async createWithSigner(
+    cometclient: CometClient,
+    signer: OfflineSigner,
+    options: SigningStargateClientOptions = {},
+  ): Promise<SigningHyperlaneModuleClient> {
+    const [account] = await signer.getAccounts();
+    return new SigningHyperlaneModuleClient(
+      cometclient,
+      signer,
+      account,
+      options,
+    );
+  }
+
+  private async submitTx<T>(
+    msg: EncodeObject,
+    options?: {
+      fee?: StdFee | 'auto' | number;
+      memo?: string;
+    },
```

Comment:
nit: would it be possible to refactor this type  into a named type so we can avoid "duplicating" it because I see the same type being used in multiple methods in this class. Could be something like this:

```typescript
type TxOptions = {
  fee?: StdFee | 'auto' | number;
  memo?: string;
}
```